src/permutation.jl

```diff
@@ -0,0 +1,162 @@
+    # see https://docs.julialang.org/en/v1/stdlib/Future/index.html
+    rnglock = ReentrantLock()
+    samp = replicate(n, use_threads=use_threads) do
+        mod = m_threads[Threads.threadid()]
+
+        local βsc = βsc_threads[Threads.threadid()]
+        local θsc = θsc_threads[Threads.threadid()]
+        lock(rnglock)
+        mod = permute!(rng, mod, blups, reterms)
+        unlock(rnglock)
+        refit!(mod)
+        (
+         objective = mod.objective,
+         σ = mod.σ,
+         β = NamedTuple{β_names}(fixef!(βsc, mod)),
+         se = SVector{p,T}(stderror!(βsc, mod)),
+         θ = SVector{k,T}(getθ!(θsc, mod)),
+        )
+    end
+    MixedModelPermutation(
+        samp,
+        deepcopy(morig.λ),
+        getfield.(morig.reterms, :inds),
+        copy(morig.optsum.lowerbd),
+        NamedTuple{Symbol.(fnames(morig))}(map(t -> (t.cnames...,), morig.reterms)),
+    )
+end
+
+function permutation(nsamp::Integer, m::LinearMixedModel; kwargs...)
+    return permutation(Random.GLOBAL_RNG, nsamp, m; kwargs...)
+end
+
+function permutation(rng::AbstractRNG, n::Integer,
+                                morig::GeneralizedLinearMixedModel;
+                                use_threads::Bool=false) where {T}
+
+    throw(ArgumentError("GLMM support is not yet implemented"))
+end
+
+permute!(mod::LinearMixedModel, blups=ranef(mod), reterms=mod.reterms) =
+    permute!(GLOBAL_RNG, mod, blups, reterms)
+
+"""
+    permute!([rng::AbstractRNG,] mod::LinearMixedModel,
+              blups=ranef(mod), reterms=mod.reterms)
+
+Simulate and install a new response via sign-flipped permutation of the residuals
+at the observational and sign-flipping of the conditional modes at group level.
+
+Sign-flipped permutation of the residuals is similar to permuting the
+(fixed-effects) design matrix. Sign-flipping the conditional modes (random effects)
+preserves the correlation structure of the random effects, while also being equivalent to
+permutation via swapped labels for categorical variables.
+
+!!! note
+    This method has serious limitations for singular models because permuting zeros
+
+See also [`permutation`](@ref), [`nonparametricbootstrap`](@ref) and [`resample!`](@ref).
+
+# Reference
+ter Braak?
+
+"""
+function permute!(rng::AbstractRNG, mod::LinearMixedModel,
+                   blups=ranef(mod),
+                   reterms=mod.reterms)
+    β = coef(mod)
+    y = response(mod) # we are now modifying the model
+    copy!(y, residuals(mod))
+
+    shuffle!(rng, y)
+    # inflate these to be on the same scale as the empirical variation instead of the MLE
+    # y .*= sdest(mod) / std(y; corrected=false)
+    # sign flipping
+    y .*= rand(rng, (-1,1), length(y))
+
+    # σ = sdest(mod)
+
+    for (re, trm) in zip(blups, reterms)
+        npreds, ngrps = size(re)
+        samp = sample(rng, 1:ngrps, ngrps; replace=false)
+
+        # use a view to avoid copying
+        newre = view(re, :, samp)
```

Comment:
random effects are permuted here. Columns are participants/items/blocks, rows are the associated predictions, so we index into that array with a permutation of columns which keeps the within-participant/item/block predictions together.